### DISCUSSION TITLE
proposal: Auto Cast for +,-,*,.... Oprators

### DISCUSSION THREAD

**Discussion by rootflood** [dotnet/csharplang#3593] [Language Ideas]
my problem https://github.com/dotnet/csharplang/issues/3572 , https://github.com/dotnet/csharplang/issues/3586
```c#
        public static void Main()
        {
            byte b = 0;

            // unexpected
            //Error Cannot implicitly convert type 'int' to 'byte'. An explicit conversion exists (are you missing a cast?)
            b = b << 3;

            // unexpected
            //Error Cannot implicitly convert type 'int' to 'byte'. An explicit conversion exists (are you missing a cast?)
            b = b + 1;

            b += 1; // no error and same to b = b + 1;
            b += 255; // no error
        }
```
the vb.net was able to same reaction in .net [Option Strict Statement](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/option-strict-statement) , [Implicit and Explicit Conversions ](https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/data-types/implicit-and-explicit-conversions) , 
```vb
    Sub M1(b As Byte)
    End Sub
    Sub M1(b As Integer)
    End Sub
    Sub M2(b As Byte)
    End Sub

    Sub Test()
        Dim b As Byte
        b = b + 5 ' this is fine. the result castes to byte
        b = b << 2 ' this is fine. the result castes to byte
        M1(b + 2) ' M1(b As Integer)
        M2(b + 2) ' M2(b As Byte)
    End Sub
```

solution :
```c#
        public static void Main()
        {
            byte b = 0;
            b = b << 3; // cast by c# to  b =(byte) (b << 3);
            b = b + 1; // cast by c# to  b =(byte) (b + 1);
        }
```

wrong cases and limits:
1- must don't cast on var keyword
```c#
        public static void Main()
        {
            byte b = 0;
            var v2 = b << 3; // it's didnt cast by c# because of var 
            // Type of v2 is int 
        }
```
2- must cast when the overloads haven't exact output of the operator type
```c#
void M1(short x) => Console.WriteLine("short");
void M1(int x) => Console.WriteLine("int");

void MV2(short x) => Console.WriteLine("short");

short b = 0;
M1(b + 1);// it's didnt cast by c# because of  " int operator +(int a, int b)" & M1(int x) output is int

MV2(b + 1);// cast by c# to  MV2((short) (b + 1));
```

answer for overflow and missing data:
1- .net have runtime overflow exception so that stop wrong cases
so for result its just same and haven't different
examples:
```c#
            byte b = 1;
            b += 255; // no runtime error
            b = 1; // no runtime error
            b =(byte) (b + 255); // no runtime error
            checked
            {
                b = 1;
                b += 255; // Overflow Exception
            }
```

2- developers knew any case of missing bits/data  
so for result its just same and haven't different
examples:
```c#
            float b = 1;
            // fine
            double D = float.MaxValue + 1;
            b +=(float) D;
            var IsSame = b == float.MaxValue; // true
```
at last please check operators and cast before comment :)
This proposal didn't change any type of oprators, it's just cast the result of operators. 
you can also check https://github.com/dotnet/csharplang/issues/2929#issuecomment-549662899 , https://github.com/dotnet/csharplang/issues/2929
if any problem seeing please make a comment, that can really help to provide best solution
My proposal must Don't destruct any .NET rules anyway
and must haven't any type of breaking changes
Also thanks @HaloFour for real helps

**Comment by theunrepentantgeek** [comment]
One of the guiding principles that the C# (and other) teams follow has the catchy name of "The Pit of Success". The idea is a simple one - that the most straightforward way of doing something should be both successful and safe. (IIRC, this phrase originated with [Scott Guthrie](https://en.wikipedia.org/wiki/Scott_Guthrie), who was head of Microsoft's Developer Division at the time.)

This principle explains a couple of choices the C# team made very early in the development of the language.

To deconstruct one of your examples, @rootflood:

``` csharp
byte b = ...
b = b << 3;
```

What is the type of subexpression `b << 3` when `b` is a **byte**?

The language specifiers had a choice: it could be a **byte** or an **int**.

Using an **int** is always safe; the result can be captured without loss regardless of the actual value in `b`.

Using a **byte** has a risk that the resulting value is too large to store in a **byte**. This leads to a further choice - should the extra bits be silently clipped or should there be an exception?

Following the principle described above, the C# language team decided that they needed to make the default behaviour *safe* so that developers would stumble into the pit of success. Since *most* developers wouldn't want to throw away data, the language keeps it by default.

The second part of the statement is the assignment `b = <intvalue>` where the lefthand can store a **byte**, the righthand contains an **int**. 

Again, the C# language team had a decision to make. Automatically narrowing the **int** was possible, but (once again) would potentially result in data loss. Experience with other langages (including, but not limited to C, C++, VB 6, and Delphi) showed the team that silently discarding data often resulted in subtle bugs. Thus, their decision was to avoid data loss, but to allow developers to opt-in to narrowing with a `(byte)` cast if they wanted.

So, this establishes why the current C# behaviour exists. It's not a bug, it didn't happen accidentally, but is the result of considered choices the team made back prior to the year 2000.

This leaves two questions.

**Is this likely to change?**

The team still uses "The Pit of Success" as a design criteria, plus they have a very strong desire for backward compatibility across versions of the language. Breaking changes do happen, but they have to cross a very high value threshold to be considered.

Introducing automatic narrowing of the value would remove the warning, bringing back all the dangers of subtle bugs that they were trying to avoid with the original decisions. This, coupled with the very high implementation cost, and with the existing straightforward syntax for a simple cast, leads me to believe that a change is highly unlikely.

**Can this possibly change?**

There's new infrastructure in the C# compiler for [target typed new](https://github.com/dotnet/csharplang/issues/100) that might just make it possible. Prior to this, the context in which the expression is used was ignored when selecting the data type. So, it might now be possible.

**Comment by rootflood** [comment]
@theunrepentantgeek 
I readed that half but I think I understand what you talking about
Yes you are right, int is safer, better for 32 bit systems and can work any where and so on....
But I didn't change operator results, 
result is always int, c# can't change that because of breaking changes and developer can have int result
So this code can explain more:
```c#
byte b = ...
b = b << 3;// b is byte 
// in fact that was :
b = (byte) (b << 3) ;
// but just cast by c# not developer
// so if developer need int he can write
int R1= b << 3;
var R2 = b << 3; // both are int
```

**Comment by AartBluestoke** [comment]
@rootflood 
byte b = (byte) (255 << 3) ; 
says 
" i am happy to discard any extra bits above the value 255"

byte b = (255 << 3) ; 
the compiler says "erm, i don't know how to fit a value bigger than 255 into a byte, can you be more specific"

most of your examples attempt to use operator (byte,int) -- in general this doesn't exist, and because byte can be promoted to int without loss, but the reverse is not true, you end up using 'int operator X (int,int)' , What you want to be using is 'byte operator X (byte,byte)'

replacing 'b' with a different variable or specific value irrespective of if it's a byte or int doesn't change how the compiler sees the operator.

"Because the shift operators are defined only for the int, uint, long, and ulong types, the result of an operation always contains at least 32 bits. If the left-hand operand is of another integral type (sbyte, byte, short, ushort, or char), its value is converted to the int type"

your byte example could be rephrased into a request for a byte sized bitshift operator.

byte b = (255 << (byte)3) ; 

your other examples work fine when passed byte:
  b = b + (byte)1;
  b += (byte)1; // no error and same to b = b + 1;
  b += (byte)255;

**Comment by rootflood** [comment]
@AartBluestoke 
Thanks for comment, but didn't talk about (byte,int) operator, in fact that was my old proposal and now I am deagree with that
This proposal doesn't declare extra operators at all. 
My proposal must Don't destruct any .NET rules anyway
and must haven't any type of breaking changes
what i want is not to be using 'byte operator + (byte,byte)'
Then what you told is right but there isn't any type of that changes
Thanks about the comment but I know oprators and I never change that in this proposal
this proposal have conflict to my other one so please don't compare that
So your told can't be right in this one, please read this proposal again
Thanks
